### ISSUE TITLE
Unable to retrive image while using the pypylon api in python3 multiprocessing

### ISSUE THREAD

**Issue by occiAdmin** [basler/pypylon#178]
I am using basler 1920-40uc. The code works fine while using in a single-threaded process. But due to the nature of the application, I have to make a process that repeatedly calls camera api and put the image data into a queue.

Below is the wrapper class that I have defined. 

`
import sys, os
from pypylon import pylon
import cv2
import logging
import time,datetime
from multiprocessing import Process

class CamProcess(Process):
    def __init__(self,queue):
        Process.__init__(self)
        self.data_queue = queue
        self.flag = True
        self.id = None
        self.cam = Basler(flag=0)
    
    def run(self):
        if self.id is None:
            self.id = os.getpid()

        while self.flag:
            ret, frame = self.cam.read()
            if ret:
                self.data_queue.put([self.id,frame])

            else:
                self.data_queue.put(None)

    def stop(self):
        self.flag = False
        self.cam.stop()

class Basler:
    def __init__ (self, flag=0,exposure=None):
        self.camera=pylon.InstantCamera (pylon.TlFactory.GetInstance ().CreateFirstDevice ())
        self.camera.Open()
        if exposure is None:
            exposure = 10000
        self.camera.ExposureTime.SetValue(int(exposure))

        if flag ==1:
            self.camera.TriggerSelector.SetValue("FrameBurstStart")
            self.camera.TriggerSource.SetValue("Line1")
            self.camera.TriggerMode.SetValue("On")

        # Grabing Continusely (video) with minimal delay
        self.camera.StartGrabbing (pylon.GrabStrategy_LatestImageOnly)
        self.converter=pylon.ImageFormatConverter ()

        # converting to opencv bgr format
        self.converter.OutputPixelFormat=pylon.PixelType_BGR8packed
        self.converter.OutputBitAlignment=pylon.OutputBitAlignment_MsbAligned
        
        
        
        """
        The code below works and it is similar to the code wrapper in 
        "read" function. Calling that function gives error
        """
        
        # self.some =self.camera.RetrieveResult (5000, pylon.TimeoutHandling_ThrowException)
        # self.image=self.converter.Convert (self.some)
        # self.img=self.image.GetArray ()
        # print(self.img)

    def read (self):

        grabResult=self.camera.RetrieveResult (10000, pylon.TimeoutHandling_ThrowException)

        if grabResult.GrabSucceeded():
            # Access the image data
            self.image=self.converter.Convert(grabResult)
            self.img=self.image.GetArray ()
            return True,self.img
        else:
            return  False, None

    def release (self):
        self.camera.StopGrabbing ()
        del self.camera

if __name__=="__main__":
    from multiprocessing import Queue
    im_q = Queue(maxsize=1)
    cam = CamProcess(im_q)
    cam.start()
    while True:
        data = im_q.get()
        print(data)
`

upon calling the function I am getting the below error.

`File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "sortingProcess.py", line 50, in run
    ret, frame = self.cam.read()
  File "/home/xyz/imaging/camera.py", line 52, in read
    ret,self.frame = self.feed.read()
  File "/home/xyz/imaging/camera.py", line 98, in read
    grabResult=self.camera.RetrieveResult (10000, pylon.TimeoutHandling_ThrowException)
  File "/home/xyz/sortingEnv/lib/python3.6/site-packages/pypylon/pylon.py", line 3299, in RetrieveResult
    return _pylon.InstantCamera_RetrieveResult(self, *args)
_genicam.RuntimeException: Error setting waitable timer 0x04487880: Invalid argument : RuntimeException thrown (file 'WaitableTimer.h', line 201)
`

**Comment by occiAdmin** [comment]
This error has taken away a good two days in a high-pressure situation. Any help is appreciated.

**Comment by thiesmoeller** [comment]
... you should instantiate the InstantCamera in the `run()` method of the `CamProcess`.

Otherwise the multiprocessing code in python will try to move th InstantCamera between the two processes. Which can't work, because this would require 
* to "pickle" the complete state of the InstantCamera 
* transfer it to the other process
* "unpickle" it
* work with it.

As you only transfer the images through your queue this shouldn't be a problem ...


```
import os
from pypylon import pylon
from multiprocessing import Process

class CamProcess(Process):
    def __init__(self, queue):
        Process.__init__(self)
        self.data_queue = queue
        self.flag = True
        self.id = None
        self.cam = None

    def run(self):
        if self.id is None:
            self.id = os.getpid()
        self.cam = Basler(flag=0)
        while self.flag:
            ret, frame = self.cam.read()
            if ret:
                self.data_queue.put([self.id, frame])

            else:
                self.data_queue.put(None)

    def stop(self):
        self.flag = False
        self.cam.stop()


class Basler:
    def __init__(self, flag=0, exposure=None):
        self.camera = pylon.InstantCamera(pylon.TlFactory.GetInstance().CreateFirstDevice())

        self.camera.Open()

        # Grabing Continously (video) with minimal delay
        self.camera.StartGrabbing(pylon.GrabStrategy_LatestImageOnly)
        self.converter = pylon.ImageFormatConverter()

        # converting to opencv bgr format
        self.converter.OutputPixelFormat = pylon.PixelType_BGR8packed
        self.converter.OutputBitAlignment = pylon.OutputBitAlignment_MsbAligned

        """
        The code below works and it is similar to the code wrapper in 
        "read" function. Calling that function gives error
        """

        # self.some =self.camera.RetrieveResult (5000, pylon.TimeoutHandling_ThrowException)
        # self.image=self.converter.Convert (self.some)
        # self.img=self.image.GetArray ()
        # print(self.img)

    def read(self):
        grab_result = self.camera.RetrieveResult(10000, pylon.TimeoutHandling_ThrowException)

        if grab_result.GrabSucceeded():
            # Access the image data
            py_image = self.converter.Convert(grab_result)
            img = py_image.GetArray()
            return True, img
        else:
            return False, None

    def release(self):
        self.camera.StopGrabbing()
        del self.camera


if __name__ == '__main__':
    from multiprocessing import Queue

    im_q = Queue(maxsize=1)
    cam = CamProcess(im_q)
    cam.start()
    while True:
        data = im_q.get()
        print(data)

```

**Comment by occiAdmin** [comment]
Thanks a lot. I couldn't comprehend  what was the reason for the error. I did test it and it is working fine.

**Comment by SarahChiu** [comment]
Hi @occiAdmin and @thiesmoeller, thanks for sharing the code and explanation! 

I tried to run the code as @thiesmoeller suggested to collect images from acA4024-29uc. The code is run in Ubuntu 20.04. 
However, the process hangs at this line in class Basler and never continues: 
```python
self.camera = pylon.InstantCamera(pylon.TlFactory.GetInstance().CreateFirstDevice())
```

When I stopped the program by Ctrl+C, it gave me this traceback: 
```bash
File "/usr/lib/python3.9/multiprocessing/popen_fork.py", line 27, in poll
  pid, sts = os.waitpid(self.pid, flag)
```
It seems like the program runs into a deadlock. Any idea why this happens?
